xwiki-platform-core/xwiki-platform-security/xwiki-platform-security-authentication/xwiki-platform-security-authentication-api/src/main/java/org/xwiki/security/authentication/UserUnauthenticatedEvent.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.xwiki.security.authentication;
+
+import org.xwiki.observation.event.Event;
+import org.xwiki.stability.Unstable;
+import org.xwiki.user.UserReference;
+
+/**
+ * Event triggered whenever a user is logged out.
+ *
+ * @version $Id$
+ * @since 16.8.0RC1
+ * @since 16.4.3
+ * @since 15.10.13
+ */
+@Unstable
+public class UserUnauthenticatedEvent implements Event
+{
+    /**
+     * The reference related to an authenticated user for whom a {@link UserUnauthenticatedEvent} has been triggered.
+     */
+    private final UserReference userReference;
+
+    /**
+     * Default constructor without user reference for matching.
+     */
+    public UserUnauthenticatedEvent()
+    {
+        this(null);
+    }
+
+    /**
+     * Default constructor.
+     *
+     * @param userReference The reference related to an authenticated user for whom a {@link UserUnauthenticatedEvent}
+     * has been triggered.
+     */
+    public UserUnauthenticatedEvent(UserReference userReference)
+    {
+        this.userReference = userReference;
+    }
+
+    /**
+     * @return the {@link UserReference} of the authenticated user.
+     */
+    public UserReference getUserReference()
+    {
+        return this.userReference;
+    }
+
+    @Override
+    public boolean matches(Object other)
+    {
+        return other instanceof UserUnauthenticatedEvent;
```

Comment:
It would worth improving this to take into account `this.userReference` as otherwise it's quite missleading to have two different constructors from the point of view of a listener.

Reply:
Honestly I'm surprised that the code for UserAuthenticatedEvent doesn't use it, since the javadoc says otherwise, see: https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-security/xwiki-platform-security-authentication/xwiki-platform-security-authentication-api/src/main/java/org/xwiki/security/authentication/UserAuthenticatedEvent.java#L39

I'd preferably have same behaviour for both events, but changing this implem in UserAuthenticatedEvent is a breaking change behaviour...

Reply:
> I'd preferably have same behaviour for both events, but changing this implem in UserAuthenticatedEvent is a breaking change behaviour...

IMO it's fixing a bug and that's typically something that should go in the abstract.

Reply:
> IMO it's fixing a bug

Would need to check usage first but I tend to agree that we should use it.

> that's typically something that should go in the abstract.

+1

Reply:
Can't find any usage in xwiki-contrib, except to trigger the event: https://github.com/search?q=org%3Axwiki-contrib%20UserAuthenticatedEvent&type=code